### PR TITLE
[Serve] Support headers in Readiness Probe

### LINKED ISSUE #3549: [Feature Request]: vLLM API-Key support for Sky Serve
Feature request for Sky Serve support of vLLMs `--api-key` flag. Presently Sky Serve has a mandatory health check in its YAML spec that cannot make unauthenticated post requests to the running vLLM server if an API key is passed. 

[See the vLLM docs here](https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html)

### PR DESCRIPTION
**Pull request by cblmemo**
Closes #3549 by adding headers configuration in readiness probe.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
     - [x] `llm/vllm/service-with-auth.yaml`
```bash
$ sky serve up llm/vllm/service-with-auth.yaml --gpus L4
Service from YAML spec: llm/vllm/service-with-auth.yaml
Service Spec:
Readiness probe method:           GET /v1/models with headers {"Authorization": "Bearer sky-authkey-3d2105b9-a9ba-4f13"}
Readiness initial delay seconds:  1200
Replica autoscaling policy:       Fixed 1 replica
Spot Policy:                      No spot policy
$ curl $(sky serve status --endpoint sky-service-4233)/v1/models -H 'Authorization: Bearer sky-authkey-3d2105b9-a9ba-4f13' 
{"object":"list","data":[{"id":"meta-llama/Llama-2-7b-chat-hf","object":"model","created":1715746959,"owned_by":"vllm","root":"meta-llama/Llama-2-7b-chat-hf","parent":null,"permission":[{"id":"modelperm-5cf910e014484663a648f66e38b08174","object":"model_permission","created":1715746959,"allow_create_engine":false,"allow_sampling":true,"allow_logprobs":true,"allow_search_indices":false,"allow_view":true,"allow_fine_tuning":false,"organization":"*","group":null,"is_blocking":false}]}]}
```
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

File: llm/vllm/service-with-auth.yaml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    # Set authorization headers here if needed.
+    headers:
+      Authorization: Bearer $AUTH_TOKEN
+  # How many replicas to manage.
+  replicas: 1
+
+# Fields below are the same with `serve-openai-api.yaml`.
+envs:
+  MODEL_NAME: meta-llama/Llama-2-7b-chat-hf
+  HF_TOKEN: <your-huggingface-token> # Change to your own huggingface token
```
Comment (Michaelvll) [inline llm/vllm/service-with-auth.yaml:16]: change this to `null` or empty as we have added #3557.

File: llm/vllm/service-with-auth.yaml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  AUTH_TOKEN: sky-authkey-3d2105b9-a9ba-4f13 # Change to your own auth token
```
Comment (Michaelvll) [inline llm/vllm/service-with-auth.yaml:17]: Maybe we should set this to `null` or empty as well.

File: sky/serve/service_spec.py [hunk 3]
```diff
@@ -220,8 +226,14 @@ def add_if_not_none(section, key, value, no_empty: bool = False):
...
         if self.post_data is None:
-            return f'GET {self.readiness_path}'
-        return f'POST {self.readiness_path} {json.dumps(self.post_data)}'
+            method = f'GET {self.readiness_path}'
+        else:
+            method = f'POST {self.readiness_path} {json.dumps(self.post_data)}'
+        if self.readiness_headers is None:
+            headers = ''
+        else:
+            headers = f' with headers {json.dumps(self.readiness_headers)}'
```
Comment (Michaelvll) [inline sky/serve/service_spec.py:235]: nit: Showing the authorization token directly in the log may not be very safe. We may want to redact the value of the headers?

**Review by Michaelvll — COMMENTED** [review]
Thanks for adding the support for the authorization token in the readiness probe @cblmemo! It looks mostly good to me!

File: llm/vllm/service-with-auth.yaml (hunk 1 above)
Comment (cblmemo) [reply llm/vllm/service-with-auth.yaml:16]: Good point! Added. Thanks!

File: sky/serve/service_spec.py (hunk 3 above)
Comment (cblmemo) [reply sky/serve/service_spec.py:235]: Good point! Changed to ` with Authotization` now.

File: sky/serve/replica_managers.py [hunk 4]
```diff
@@ -567,7 +570,8 @@ def __init__(self, service_name: str,
         self._update_mode = serve_utils.DEFAULT_UPDATE_MODE
         logger.info(f'Readiness probe path: {spec.readiness_path}\n'
                     f'Initial delay seconds: {spec.initial_delay_seconds}\n'
-                    f'Post data: {spec.post_data}')
+                    f'Post data: {spec.post_data}\n'
+                    f'Readiness headers: {spec.readiness_headers}')
```
Comment (Michaelvll) [inline sky/serve/replica_managers.py:574]: Seems we are still printing out the headers? Should we use `list(spec.readiness_headers.keys())` instead?

**Review by Michaelvll — APPROVED** [review]
Thanks for adding this @cblmemo! LGTM with the comment fixed.

File: sky/serve/replica_managers.py (hunk 4 above)
Comment (cblmemo) [reply sky/serve/replica_managers.py]: Good catch! Fixed. Thanks!

**Pull request merged**